### PR TITLE
3473 virtualenv advanced concept

### LINKED ISSUE #3393: VirtualENV Advanced Concept - Documentation
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3473

Originally created by @vischous on 2022-05-09 19:46:27

---

Add a VirtualENV Advanced Concept Page. Created from the discussion here https://gitlab.com/meltano/meltano/-/merge_requests/2638#note_940235684

### PR DESCRIPTION
**Pull request by visch**
Closes https://github.com/meltano/meltano/issues/3393, and https://github.com/meltano/meltano/issues/5913

Old issue https://gitlab.com/meltano/meltano/-/issues/3473 

> _Merges 3473-virtualenv-advanced-concept -> master_
> 
> Migrated from GitLab: https://gitlab.com/meltano/meltano/-/merge_requests/2640
> 

* [x]  Include the proposed fix or feature
* [NA] Include and update tests for the modified code
* [x]  Include a documentation change
* [x]  Add a [CHANGELOG.md](https://gitlab.com/meltano/meltano/blob/master/README.md#Changelog) entry in the Unreleased section for any user-facing changes
* [x]  Mention or assign a maintainer 
Closes https://gitlab.com/meltano/meltano/-/issues/3473

### CODE REVIEW

**Review by tayloramurphy — APPROVED** [review]
Pushed some changes - I'm pretty happy with this from a docs perspective.

**Comment by tayloramurphy** [comment]
@aaronsteers will you have an engineer review this for accuracy around all things python?

File: docs/src/_concepts/python_virtual_environments.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+```
+
+That's it! You've created a virtual environment. Feel free to explore the directory and compare it to your global python directory!
+
+## Activating Your Virtual Environment
+
+Activate the virtual environment, and upgrade pip using:
+
+<div class="notification is-info">
+  <p>Note that pip needs to be upgraded everytime you make a new venv, you want to do this to avoid hard to troubleshoot issues with dependencies later.</p>
```
Comment (WillDaSilva) [inline docs/src/_concepts/python_virtual_environments.md:49]:
```suggestion
  <p>Note that pip needs to be upgraded every time you make a new venv. Doing this helps avoid hard to troubleshoot issues with dependencies later.</p>
```

File: docs/src/_concepts/python_virtual_environments.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+We're going to focus on the Meltano use case for Virtual Environments which is use for a two main purposes
+1. Installing Meltano in an isolated environment
+1. Installing Plugins (Taps, Targets, Transformers, etc) in isolated environments
+
+# Why should I care about Virtual Environments?
+
+Ideally you don't have to to worry about Virtual Environments while using Meltano, 
+we recommend using pipx to install meltano which manages creating venvs for you, see our [Installation Guide](../_guide/installation.md).
+
+However, if you ever need to customize or build your own production pipeline (Or do anything else) you may need to understand
```
Comment (WillDaSilva) [inline docs/src/_concepts/python_virtual_environments.md:25]:
```suggestion
However, if you ever need to customize or build your own production pipeline (or do anything else) you may need to understand
```

File: docs/src/_concepts/python_virtual_environments.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  <p>Note that pip needs to be upgraded everytime you make a new venv, you want to do this to avoid hard to troubleshoot issues with dependencies later.</p>
+</div>
+
+```bash
+source .venv/meltano/bin/activate
+pip install --upgrade pip
+```
+
+If the virtual environment was activated successfully, you'll see a `(meltano)` indicator added to
+your prompt.
```
Comment (WillDaSilva) [inline docs/src/_concepts/python_virtual_environments.md:58]: This behaviour was deprecated, and has been removed from newer versions of `venv`.
```suggestion
```

File: docs/src/_concepts/python_virtual_environments.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+To streamline this process, you can define a [shell alias](https://shapeshed.com/unix-alias/)
+that'll be easier to remember than the entire activation invocation:
+
+```bash
+# add to `~/.bashrc`, `~/.zshrc`, etc, depending on the shell you use:
+alias meltano!="source $MELTANO_PROJECT_PATH/.venv/meltano/bin/activate"
+ 
+# use as follows, after creating a new shell:
+meltano!
+```
```
Comment (WillDaSilva) [inline docs/src/_concepts/python_virtual_environments.md:73]:
```suggestion
```

Users who are familiar with this sort of stuff won't need instructions on how to create a shell alias/function as a shortcut. Users who are not familiar with this sort of stuff would probably be better off with less information being presented to them. When in doubt, keep it simple.

File: docs/src/_concepts/python_virtual_environments.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+You can deactivate a virtual environment by typing `deactivate` in your shell.
+
+### Install Meltano into a Virtual Environment
+
+Now that you have your virtual environment set up and running, run the following command to install
+the Meltano package:
+
+```bash
+pip3 install meltano
```
Comment (WillDaSilva) [inline docs/src/_concepts/python_virtual_environments.md:83]:
```suggestion
pip install meltano
```

File: docs/src/_concepts/python_virtual_environments.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+```
+
+# How does meltano use Virtual Environments internally?
+
+Whenever you run `meltano install`, Meltano creates a `.meltano/` directory in your project.
+This folder has a number of sub folders (these folders are subject to change at any point as this is an internal folder).
+In `.meltano/` you may see an `extractors/` folder that was created (if you have an extractor). 
+Go into that folder, you'll see the name of a your tap, inside of that folder is a virtual environment!
```
Comment (WillDaSilva) [inline docs/src/_concepts/python_virtual_environments.md:91]:
```suggestion
This directory has a number of sub-directories (subject to change at any point, as this is an
internal directory). In `.meltano/` you may see an `extractors/` directory that was created (if you
have an extractor). Go into that directory, you'll see the name of a your tap, inside of that
directory is a virtual environment!
```

**Review by WillDaSilva — APPROVED** [review]

File: docs/src/_concepts/python_virtual_environments.md (hunk 3 above)
Comment (visch) [reply docs/src/_concepts/python_virtual_environments.md:58]: I'm going to remove the /meltano folder then, you have some crazy good insight I had no idea :D

Comment (WillDaSilva) [reply docs/src/_concepts/python_virtual_environments.md]: Ah, my bad. It was deprecated and removed from virtual envs managed by `pyenv`, not by `venv`.

**Comment by tayloramurphy** [comment]
Oh it seems like my changes never actually got pushed :(

**Comment by tayloramurphy** [comment]
@visch evidently I don't know what I'm doing. Can you bring in the changes I have in https://github.com/meltano/meltano/commit/f531fa6310d1d67e55cf8c941e4224763ccfaf01 into this PR?

**Comment by tayloramurphy** [comment]
@visch amazing! Merging 🚢

**Pull request merged**